### PR TITLE
Allow Pixmap to be backed by an unowned slice

### PR DESCRIPTION
**Pull request by nilclass**
First of all, thank you for a great drawing library. I've tried a few alternatives, and this seems a very promising project.

Something I found lacking though is the ability to draw onto a surface that is not managed by the library itself, like cairo's [`cairo_image_surface_create_for_data`](https://cairographics.org/manual/cairo-Image-Surfaces.html#cairo-image-surface-create-for-data) does.

The changes in this PR are my attempt to add that ability.
The basic idea is that a `Pixmap` can be backed by either a `Vec<u8>` (which it owns) or by a `&mut [u8]` slice (which it does not own). A `Pixmap` with unowned data can be created with:
```rust
let pixmap = Pixmap::from_data(800, 600, &mut my_u8_slice);
```

There are a number of reasons why this can be useful, such as:
- Drawing to a [framebuffer device](https://en.wikipedia.org/wiki/Linux_framebuffer), which usually works by mapping the `/dev/fbX` device into memory using `mmap` and then drawing to the mapped region.
- Drawing to shared memory. I've added an example for this, which is a bit contrived. A real world use-case would be writing applications for wayland, which [uses shared memory pools](https://wayland-book.com/surfaces/shared-memory.html) to let clients draw onto surfaces.
- Mixing tiny-skia with other drawing libraries (such as for text rendering, as you mention in #16).

### CODE REVIEW

**Comment by RazrFalcon** [comment]
Yes, this is an important feature, but I'm still not sure how to implement it. Your implementation is fine, but it assumes that provided buffer is premultiplied RGBA, which is not always the case.

**Comment by nilclass** [comment]
True. For dealing with buffer pools, supporting a `stride` different from `width` would also be useful, as well as accepting buffers that are larger than the size of the pixmap.

Regarding the pixel format, the first thing that comes to mind would be making `Pixmap` generic with respect to the pixel type instead of using `PremultipliedColorU8` directly. I can't judge how feasible that is though.
It would also require some conversions, when dealing with more than one pixmap, like in a `Pattern` or in `Canvas.draw_pixmap`.
However, within the README you wrote that "Non-RGBA8888 images" are out-of-scope. Do you plan to stick with that?

**Comment by RazrFalcon** [comment]
I would need some time to think about it. The raster pipeline itself is u16/f32 RGBA. We can't change that. So we have to do what Skia does - convert the provided pixmap data to the internal format. Ideally, we should have Pixmap as a trait, which defines X to RGBA8/RGBAF32 conversion. And then we can have Pixmap type, that is just a default and the only implementation in the tiny-skia itself.

The main problem here is that the caller have to write SIMD for conversions by itself.

So the main challenge here is to have a nice, universal pixmap abstraction trait that doesn't affect performance that much.

Something like:

```
trait Pixmap {
    fn to_rgba_f32(&mut self, x: u32, y: u32, buf: &mut [f32; 8]);
    fn to_rgba_u16(&mut self, x: u32, y: u32, buf: &mut [u16; 16]);
    fn is_premultiplied(&self) -> bool; // or const?
}
```

Basically, I want this feature, but I don't want it to be a tiny-skia problem, but rather a caller problem.

PS: afaik, Skia doesn't support stride at all.

**Comment by RazrFalcon** [comment]
Also, this means that `Canvas` would become `Canvas<T: Pixmap>`.

**Comment by RazrFalcon** [comment]
I think I would implement it myself, because it's a very low-level feature.

**Comment by RazrFalcon** [comment]
So I've tried a trait approach and it either 4x slower or ultra-low level. Looks like it has to be implemented in the library itself. And we basically have to replicate Skia logic, which would take some time.

**Comment by RazrFalcon** [comment]
I've published my implementation. It has a more radical impact on the API, but the idea is roughly the same.

I still not sure what to do with non RGBA8 images. It's a very performance-sensitive topic. For now, I would say that premultiplying/demultiplying and byte-swapping (RGBA<->BGRA<->ABGR) is up to the caller. From the CPU standpoint, it's way faster to do this once in a hot loop than doing on each load/store in the pipeline.

**Comment by nilclass** [comment]
Nice! I like about your solution that the source in `draw_pixmap` doesn't require a mutable slice.

One minor issue though: `PixmapMut::from_bytes` / `PixmapRef::from_bytes` take an `IntSize`, which is not a public type.

**Comment by RazrFalcon** [comment]
>draw_pixmap doesn't require a mutable slice.

Yes. Same with `Pattern`.

>take an IntSize, which is not a public type

My bad. Will fix it now.
